### ISSUE TITLE
Increase stack size for host when running in IIS In-Proc

### ISSUE THREAD

**Issue by anurse** [aspnet/AspNetCore#9394]
Epic https://github.com/aspnet/AspNetCore/issues/8833

Blocked by https://github.com/dotnet/coreclr/issues/21450

Once there's an API/Property to increase the stack size, we need to react to it in IIS in two ways:

1. Increase the default to something more appropriate for our scenario
2. Make it configurable in web.config (if possible).

**Comment by anurse** [comment]
ETA of the blocking issue is preview6.

**Comment by Projkt-James** [comment]
@anurse can the stack size limits cause issues with the inbuilt Dependency Injection or am I more likely having a dependency recursion problem? I'm having an overflow exception only present when moving to InProcess hosting.

**Comment by anurse** [comment]
I'm not specifically aware, but it is possible. The stack size **when running in IIS** is significantly smaller than when running as a standard .NET Console app. The easiest way to test is to run your app directly on Kestrel and see if you still hit the `StackOverflowException`. If you do, it's more likely to be a runaway recursive call, if you don't it may just be a deep stack that's hitting the IIS limit.

**Comment by Projkt-James** [comment]
@anurse thank you for the timely response, I have no issues when running on Kestrel, hopefully this change in 3.0 fixes it up.

**Comment by anurse** [comment]
Good to know! Yeah, that indicates you should have a better experience when we get this change in.

**Comment by anurse** [comment]
Need to review if the corefx API is present.

**Comment by martincostello** [comment]
@Projkt-James We seemed to experience the issue with 2.2.x when doing model binding, particularly when re-executing a request in an error handler.

**Comment by anurse** [comment]
Yep, I believe model binding can also produce fairly deep stacks. This should be improved with this coming change.

**Comment by jkotalik** [comment]
https://github.com/dotnet/core-setup/pull/5859 should be merge into our repo soon, so we can work on this.

**Comment by anurse** [comment]
Still blocked on https://github.com/dotnet/coreclr/issues/21450

**Comment by anurse** [comment]
CoreCLR builds starting with https://github.com/dotnet/coreclr/commit/6a85ca560a7dfa109c6f17c0b76d93373dbd5a00 now support the `COMPlus_DefaultStackSize` environment variable for configuring the default stack size in bytes (see https://github.com/dotnet/coreclr/pull/24532).

There is also a runtime property `DEFAULT_STACK_SIZE`, but configuring that in ANCM is blocked at the moment (#9520)

**Comment by anurse** [comment]
Apparently we're behind on deps here, waiting on a new coreclr. BAR says we're 7 days old from coreclr and the commit referenced above is 7 days old.

**Comment by jkotalik** [comment]
**Acceptance checklist** (check one item)
* [ ] We decided not to take this fix.
* [ ] The fix is tests-only.
* [x] The fix contains product changes (**check all items below**).
    * [ ] Relevant XML documentation comments for new public APIs are present.
    * [x] Narrative docs (docs.microsoft.com) are updated. (**check one item below**)
        * [ ] The change requires a new article. An issue with an outline has been filed here: [ISSUE LINK]
        * [x] The change requires a change to an existing article. A **closed** docs PR with these changes is linked here: https://github.com/aspnet/AspNetCore.Docs/pull/12590 (note, PR is still open)
        * [ ] The change requires no docs changes.
    * [x] Verification has been completed. (**check one item below**)
        * [x] The change is in the shared framework and was verified against the following versions 
            * SDK installer: 3.0.100-preview6-012113
            * ASP.NET Core Runtime: 3.0.0-preview6-19279-01
            * .NET Core Runtime: 3.0.0-preview6-27728-04
        * [ ] The change is in an OOB NuGet/NPM/JAR package and was verified against the following version of that package: [PACKAGE ID] [VERSION]
